### PR TITLE
Automatically infer and expose used tables from used gates

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR
* tracks the table associated with each gate that uses lookups
* generates the table when used, automatically inserting it into the list of tables considered when creating the proof
* removes the explicit use of `xor_table` in the chacha test

### CODE REVIEW

File: kimchi/src/circuits/gate.rs [hunk 1]
```diff
@@ -235,6 +235,7 @@ impl<F: FftField> LookupInfo<F> {
 
     /// Each entry in `kinds` has a corresponding selector polynomial that controls whether that
     /// lookup kind should be enforced at a given row. This computes those selector polynomials.
+    #[allow(clippy::type_complexity)]
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:238]: Is there a way to make these returned types clearer? I think Clippy is right to warn here.

File: kimchi/src/circuits/gate.rs [hunk 2]
```diff
@@ -189,16 +182,17 @@ pub enum LookupsUsed {
 impl<F: FftField> LookupInfo<F> {
     /// Create the default lookup configuration.
     pub fn create() -> Self {
-        let kinds = lookup_kinds::<F>();
+        let (kinds, kinds_set): (Vec<_>, Vec<_>) =
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs]: The type signature seems unecessary

File: kimchi/src/circuits/gate.rs [hunk 3]
```diff
@@ -189,16 +182,17 @@ pub enum LookupsUsed {
...
         LookupInfo {
             max_joint_size: kinds.iter().fold(0, |acc0, v| {
                 v.iter()
                     .fold(acc0, |acc, j| std::cmp::max(acc, j.entry.len()))
             }),
 
-            kinds_map: GateType::lookup_kinds_map::<F>(),
-            kinds,
+            kinds_map: GateType::lookup_kinds_map::<F>(kinds_set),
+            kinds: kinds,
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs]: `kinds,` is enough

File: kimchi/src/circuits/gate.rs [hunk 4]
```diff
@@ -380,24 +398,31 @@ impl GateType {
...
+                locations_with_tables.push(GatesLookupSpec {
+                    gate_positions: locations,
+                    gate_lookup_table: table,
+                });
             }
             (patterns, locations_with_tables)
         }
     }
 
     #[allow(clippy::type_complexity)]
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:412]: You forgot to remove this :)

File: kimchi/src/circuits/gate.rs [hunk 5]
```diff
@@ -342,25 +385,62 @@ impl GateType {
...
     }
 
-    pub fn lookup_kinds_map<F: Field>() -> HashMap<(GateType, CurrOrNext), usize> {
-        let mut res = HashMap::new();
-        let lookup_kinds = Self::lookup_kinds::<F>();
-        for (i, (_, locs)) in lookup_kinds.into_iter().enumerate() {
-            for (g, r) in locs {
-                if let std::collections::hash_map::Entry::Vacant(e) = res.entry((g, r)) {
+    #[allow(clippy::type_complexity)]
+    pub fn lookup_kinds_map<F: Field>(
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:413]: What does function do? Could use some doc

File: kimchi/src/circuits/gate.rs [hunk 6]
```diff
@@ -273,6 +298,24 @@ impl<F: FftField> LookupInfo<F> {
     }
 }
 
+/// Specifies the relative position of gates and the fixed lookup table (if applicable) that a
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:301]: No biggy, but I always try to line break after punctuation and not in the middle of a sentence as it looks weird otherwise

File: kimchi/src/circuits/gate.rs [hunk 7]
```diff
@@ -273,6 +298,24 @@ impl<F: FftField> LookupInfo<F> {
...
+/// given lookup configuration should apply to.
+pub struct GatesLookupSpec {
+    /// The set of positions relative to an active gate where a lookup configuration applies.
+    pub gate_positions: HashSet<(GateType, CurrOrNext)>,
+    /// The fixed lookup table that should be used for these lookups, if applicable.
+    pub gate_lookup_table: Option<GateLookupTable>,
+}
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:308]: What are cases when there is no fixed lookup table associated?

File: kimchi/src/circuits/gate.rs [hunk 8]
```diff
@@ -273,6 +298,24 @@ impl<F: FftField> LookupInfo<F> {
...
+
+/// Specifies mapping from positions defined relative to gates into lookup data.
+pub struct GatesLookupMaps {
+    /// Enumerates the selector that should be active for a particular gate-relative position.
+    pub gate_selector_map: HashMap<(GateType, CurrOrNext), usize>,
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:313]: Should that usize be an enum?

File: kimchi/src/circuits/gate.rs [hunk 9]
```diff
@@ -273,6 +298,24 @@ impl<F: FftField> LookupInfo<F> {
...
+    /// Enumerates the fixed tables that should be used for lookups in a particular gate-relative
+    /// position.
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:315]: No need to break that comment for a single word

File: kimchi/src/circuits/gate.rs [hunk 10]
```diff
@@ -186,18 +182,35 @@ pub enum LookupsUsed {
     Joint,
 }
 
+/// Enumerates the different 'fixed' lookup tables used by individual gates
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub enum GateLookupTable {
+    Xor,
+}
+
+pub fn get_table<F: FftField>(table_name: GateLookupTable) -> Vec<Vec<F>> {
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:191]: Missing doc on this and the previous enum

File: kimchi/src/circuits/gate.rs [hunk 11]
```diff
@@ -342,25 +385,62 @@ impl GateType {
...
+            let mut patterns = Vec::with_capacity(lookups.len());
+            let mut locations_with_tables = Vec::with_capacity(lookups.len());
+            for (pattern, locations, table) in lookups {
+                patterns.push(pattern);
+                locations_with_tables.push(GatesLookupSpec {
+                    gate_positions: locations,
+                    gate_lookup_table: table,
+                });
+            }
+            (patterns, locations_with_tables)
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:408]: There should be a type for patterns, it’s too chaotic imo to have all these tuples. 

```rust
Struct Pattern {
  /// what lookups to perform in a row
  Lookups: Vec<JointLookup>
  /// the gate that has these lookups
  Locations: HashSet<(GateType,  …)>
  /// the table used in the lookups
  Table
}
```

Comment (mimoo) [reply kimchi/src/circuits/gate.rs:408]: I’m also wondering if this is the best way to specify the layout of all these lookups. This might be unrelated to your PR, I still find it extremely hard to follow these values and how they are loosely related

Comment (mimoo) [reply kimchi/src/circuits/gate.rs:408]: Imo we should build some sort of abstractions that gate implementors can then use to construct fixed tables, and use table lookups as part of their gates. If the abstraction is clean, it’ll be easier to build on top of that in the future. Wdyt?

File: kimchi/src/circuits/gate.rs [hunk 12]
```diff
@@ -342,25 +385,62 @@ impl GateType {
...
+        for (
+            i,
+            GatesLookupSpec {
+                gate_positions: locs,
+                gate_lookup_table: table_kind,
+            },
+        ) in locations_with_tables.into_iter().enumerate()
+        {
+            for location in locs {
+                if let std::collections::hash_map::Entry::Vacant(e) = index_map.entry(location) {
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:427]: Import Entry instead

File: kimchi/src/circuits/gate.rs [hunk 13]
```diff
@@ -342,25 +385,62 @@ impl GateType {
...
+                    if let std::collections::hash_map::Entry::Vacant(e) = table_map.entry(location)
+                    {
+                        e.insert(table_kind);
+                    }
+                }
             }
         }
-        res
+        GatesLookupMaps {
+            gate_selector_map: index_map,
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:441]: Index_map should be called gate_selector_map throughout, same for table_map

File: kimchi/src/circuits/gate.rs [hunk 14]
```diff
@@ -342,25 +385,62 @@ impl GateType {
...
+            for location in locs {
+                if let std::collections::hash_map::Entry::Vacant(e) = index_map.entry(location) {
                     e.insert(i);
                 } else {
                     panic!("Multiple lookup patterns asserted on same row.")
                 }
+                if let Some(table_kind) = table_kind {
+                    if let std::collections::hash_map::Entry::Vacant(e) = table_map.entry(location)
+                    {
+                        e.insert(table_kind);
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:435]: I feel like “kind” is used differently in the lookup code, but maybe it’s me misunderstanding the code. It’s also called table_id elsewhere

File: kimchi/src/circuits/gate.rs [hunk 15]
```diff
@@ -225,35 +238,47 @@ impl<F: FftField> LookupInfo<F> {
...
+        let mut gate_tables = HashSet::new();
 
         // TODO: is take(n) useful here? I don't see why we need this
         for (i, gate) in gates.iter().enumerate().take(n) {
             let typ = gate.typ;
 
-            if let Some(v) = self.kinds_map.get(&(typ, CurrOrNext::Curr)) {
-                res[*v][i] = F::one();
+            if let Some(selector_index) = self.kinds_map.get(&(typ, CurrOrNext::Curr)) {
+                selector_values[*selector_index][i] = F::one();
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:256]: I’m not a fan of this selector index, but confused enough about the flow that I wouldn’t be able to suggest a better way

File: kimchi/src/circuits/gate.rs [hunk 16]
```diff
@@ -225,35 +238,47 @@ impl<F: FftField> LookupInfo<F> {
...
     /// Each entry in `kinds` has a corresponding selector polynomial that controls whether that
     /// lookup kind should be enforced at a given row. This computes those selector polynomials.
-    pub fn selector_polynomials(
+    #[allow(clippy::type_complexity)]
+    pub fn selector_polynomials_and_tables(
         &self,
         domain: &EvaluationDomains<F>,
         gates: &[CircuitGate<F>],
-    ) -> Vec<E<F, D<F>>> {
+    ) -> (Vec<E<F, D<F>>>, Vec<Vec<Vec<F>>>) {
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:246]: If I understand correctly, this vec of vec of vec is simply a list of all tables before being combined. We should have a type for a table (and a type for a combined table)

File: kimchi/src/circuits/gate.rs [hunk 17]
```diff
@@ -186,18 +182,35 @@ pub enum LookupsUsed {
...
+    match table_name {
+        GateLookupTable::Xor => crate::circuits::polynomials::chacha::xor_table(),
+    }
+}
+
 impl<F: FftField> LookupInfo<F> {
     /// Create the default lookup configuration.
     pub fn create() -> Self {
-        let kinds = lookup_kinds::<F>();
+        let (kinds, locations_with_tables): (Vec<_>, Vec<_>) = GateType::lookup_kinds::<F>();
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:200]: Unecessary type sig

**Review by mimoo — APPROVED** [review]
I spent several hours looking at the code and I’m still very confused about the flow. I think the abstractions are leaking, and I’m really worried that if we continue to build on the current implementation it’ll become harder and harder to reason about the logic. That being said these are mostly comments unrelated to the PR and I this looks like a step in the right direction to me.

File: kimchi/src/circuits/constraints.rs [hunk 18]
```diff
@@ -331,17 +331,25 @@ impl<F: FftField + SquareRootField> LookupConstraintSystem<F> {
...
                 let d1_size = domain.d1.size();
 
+                let (lookup_selectors, gate_lookup_tables) =
+                    lookup_info.selector_polynomials_and_tables(domain, gates);
+
                 // get the last entry in each column of each table
-                let dummy_lookup_values: Vec<Vec<F>> = lookup_tables
+                let dummy_lookup_values: Vec<Vec<F>> = gate_lookup_tables
                     .iter()
+                    .chain(lookup_tables.iter())
```
Comment (mimoo) [inline kimchi/src/circuits/constraints.rs:340]: So these are non-fixed lookup tables?

Comment (mimoo) [reply kimchi/src/circuits/constraints.rs:340]: in which case, I would suggest making that argument an `Option`, and also does that work currently or is that a step in the direction of making it work?

File: kimchi/src/circuits/gate.rs (hunk 7 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:308]: When you're using a runtime or side-loaded lookup table. So none yet, but this is future-proofing.

File: kimchi/src/circuits/gate.rs (hunk 8 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:313]: I don't think so, since the number and order of the selectors changes depending on the gates that are active in a circuit.

File: kimchi/src/circuits/gate.rs (hunk 15 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:256]: My ultimate plan is for them to go away entirely. IMO we should just be using the gate selectors (added together, under the assumption that they're mutually exclusive) rather than creating a dedicated selector.

File: kimchi/src/circuits/gate.rs (hunk 14 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:435]: `table_id` is a distinct, but related, concept. I don't think we should confuse the two here.

File: kimchi/src/circuits/gate.rs (hunk 13 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:441]: We'd still need somewhere for the map to live if we want to use it in a closure. Might as well store it here directly.

File: kimchi/src/circuits/gate.rs (hunk 6 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:304]: I'm already of the opinion that the comments in rust are too long (they're a nuisance to read on a phone, for example), I'd prefer not to make this line even longer.

File: kimchi/src/circuits/gate.rs (hunk 9 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:318]: As above, the line is already too long IMO

File: kimchi/src/circuits/gate.rs (hunk 10 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs]: The name already says what it does; I don't know what comment I could add that wouldn't be redundant.

File: kimchi/src/circuits/gate.rs (hunk 14 above)
Comment (mimoo) [reply kimchi/src/circuits/gate.rs:435]: I was pointing to that also:

```rust
pub struct LookupInfo<F> {
    /// A single lookup constraint is a vector of lookup constraints to be applied at a row.
    /// This is a vector of all the kinds of lookup constraints in this configuration.
    pub kinds: Vec<Vec<JointLookup<F>>>,
```

which seems to be a different "kind"

File: kimchi/src/circuits/gate.rs (hunk 11 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:410]: > What do we really want to get here?
> 
> * Kimchi can have gates that can perform lookups, oddly gates actually don’t encode these, they are encoded outside of the gate itself.

This is kinda inverted; some gates 'require' lookups in order to behave correctly, but the gates themselves do not perform them. I agree that it would be better if this lived alongside the gates.

> * Then a gate can perform lookups on two rows at a time, which is opaque from the outside (and might lead to UB if the next row does something weird)

Again, the gate isn't doing it. There's an assertion in this file that ensures that we don't trigger 2 lookup configurations on the same line.

> * A gate can also perform several lookups, but they all have to be for the same table. Oddly we don’t always carry this information

They don't have to be from the same table. That they currently are simplified things for this PR, but nothing about the lookup's setup cares about the particular table.

> * A lookup is either a single lookup (you want to check for the existence of a special value) or for a joint lookup (your table has more than one column). Oddly, a jointlookup is built on top of a single lookup

We never use single lookup in isolation, and once we add `table_id` properly, we probably never will. We should rename them to something that reflects that 'single' is a linear combination of witness fields to form one value in the lookup, and that 'joint' is the actual entry in the table, comprising one or more 'single' lookups and the table ID thereof.

> * A single lookup, and thus a joint lookup, can be formed from a linear combination of witness columns (not used currently)

I take it you haven't looked at `chacha_final_pattern`...

> Imo we should build some sort of abstractions that gate implementors can then use to construct fixed tables, and use table lookups as part of their gates. If the abstraction is clean, it’ll be easier to build on top of that in the future. Wdyt? I’m scared that if we continue to build on top of the current abstractions the logic will become harder and harder to work on.

I'm going to try and improve things as I work through implementing 'extended lookup tables', but my priority is getting that done rather than iterating on the existing one.

File: kimchi/src/circuits/gate.rs (hunk 5 above)
Comment (mimoo) [reply kimchi/src/circuits/gate.rs:414]: every public function should be documented, please add one here

File: kimchi/src/circuits/gate.rs (hunk 6 above)
Comment (mimoo) [reply kimchi/src/circuits/gate.rs:304]: These are really meant to be read on the generated docs though

File: kimchi/src/circuits/gate.rs (hunk 9 above)
Comment (mimoo) [reply kimchi/src/circuits/gate.rs:318]: that sounds like ideology :P

File: kimchi/src/circuits/gate.rs (hunk 10 above)
Comment (mimoo) [reply kimchi/src/circuits/gate.rs]: as above, public functions need to be documented, even if redundant

File: kimchi/src/circuits/constraints.rs (hunk 18 above)
Comment (mrmr1993) [reply kimchi/src/circuits/constraints.rs:340]: IMO an option is redundant when we're passing a slice/vector, since `None` should be equivalent to the empty slice (and indeed `Vec::with_capacity(0)` is non-allocating, so it's a drop-in replacement).

File: kimchi/src/circuits/gate.rs [hunk 19]
```diff
@@ -188,7 +188,9 @@ pub enum GateLookupTable {
     Xor,
 }
 
-pub fn get_table<F: FftField>(table_name: GateLookupTable) -> Vec<Vec<F>> {
+pub type LookupTable<F> = Vec<Vec<F>>;
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:193]: missing doc :P
